File: stack.cpp
#include <iostream>

template<typename T>
class Stack {
private:
    T* stackArray;
    int top;
    int maxSize;

public:
    
    Stack(int size) {
        stackArray = new T[size];
        maxSize = size;
        top = -1;
    }

    ~Stack() {
        delete[] stackArray;
    }

    
    void push(T item) {
        if (top >= maxSize - 1) {
            std::cout << "Stack overflow" << std::endl;
        } else {
            stackArray[++top] = item;
        }
    }

    
    T pop() {
        if (top < 0) {
            std::cout << "Stack underflow" << std::endl;
            return T();  
        } else {
            return stackArray[top--];
        }
    }

   
    T topElement() const {
        if (top < 0) {
            std::cout << "Stack is empty" << std::endl;
            return T();
        } else {
            return stackArray[top];
        }
    }

    
    bool isEmpty() const {
        return top < 0;
    }

   
    int size() const {
        return top + 1;
    }
};

int main() {
    
    Stack<int> intStack(5);
    intStack.push(10);
    intStack.push(20);
    intStack.push(30);

    std::cout << "Top (int): " << intStack.topElement() << std::endl;  
    std::cout << "Pop (int): " << intStack.pop() << std::endl;         
    std::cout << "Size (int): " << intStack.size() << std::endl;       
    
    Stack<std::string> stringStack(5);
    stringStack.push("Hello");
    stringStack.push("World");

    std::cout << "Top (string): " << stringStack.topElement() << std::endl;  
    std::cout << "Pop (string): " << stringStack.pop() << std::endl;         
    std::cout << "Size (string): " << stringStack.size() << std::endl;       

    return 0;
}

File: stack.py
class Stack:
    def __init__(self):
        self.stack = []

    def push(self, item):
        
        self.stack.append(item)

    def pop(self):
       
        if not self.is_empty():
            return self.stack.pop()
        else:
            return "Stack is empty"

    def peek(self):
        """스택의 가장 위에 있는 항목을 반환"""
        if not self.is_empty():
            return self.stack[-1]
        else:
            return "Stack is empty"

    def is_empty(self):
       
        return len(self.stack) == 0
    
    def is_empty(self):
        
        if len(self.stack) == 0:
            return True
        else:
            return False
        
        

    def size(self):
        
        return len(self.stack)

    def __str__(self):
        
        return str(self.stack)

# 스택 사용 예제
my_stack = Stack()
my_stack.push(1)
my_stack.push(2)
my_stack.push(3)

print("Stack:", my_stack)  # [1, 2, 3]
print("Stack Size:", my_stack.size())  # 3

print("Pop:", my_stack.pop())  # 3
print("Stack Size:", my_stack.size())  # 2

print("Stack Peek:", my_stack.peek())  # 2
print("Stack isEmpty:", my_stack.is_empty())  # False